File: algorithms/palindrome-number/palindrome.js
// https://leetcode.com/problems/palindrome-number/

// Best
// Time - O(n) -- O(log10(n))
// Space - O(1)
var isPalindrome = function (x) {
  if (x < 0) return false
  let reversed = 0
  let remainder
  let original = x

  while (x > 0) {
    remainder = x % 10
    reversed = reversed * 10 + remainder
    x = Math.floor(x / 10)

  }
  return original == reversed
};

// Alternative - String method
function splitNum(num) {
  return String(num).split("").map(Number);
}

var isPalindrome = function (x) {
  if (x < 0) return false

  const arr = splitNum(x)
  const iterArr = arr.length - 1
  let n = 0
  while (true) {
    if ((arr[n] ^ arr[iterArr - n]) === 0) {
      if (n >= iterArr) {
        return true
      }
      n++
      continue
    } else {
      return false
    }
  }
};

File: algorithms/palindrome-number/palindrome.cpp
// https://leetcode.com/problems/palindrome-number/

// Best
// Time - O(n) -- O(log10(n))
// Space - O(1)
class Solution
{
public:
  bool isPalindrome(int x)
  {
    int original = x;
    if (x < 0)
      return false;

    long long reversed = 0;
    while (x >= 10)
    {
      reversed = reversed * 10 + x % 10;
      x /= 10;
    }
    reversed = reversed * 10 + x;
    return reversed == original;
  }
};

// Similar, much concise
class Solution
{
public:
  bool isPalindrome(int x)
  {
    int original = x;
    if (x < 0)
      return false;

    long long reversed = 0;
    while (x > 0)
    {
      reversed = reversed * 10 + x % 10;
      x /= 10;
    }

    return reversed == original;
  }
};